padertorch/data/segment.py

```diff
@@ -0,0 +1,442 @@
+from copy import copy
+from typing import Union
+
+import numpy as np
+from paderbox.array import segment_axis
+from paderbox.utils.nested import flatten, deflatten
+from padertorch.utils import to_list
+
+possible_anchor_modes = [
+    'left',
+    'right',
+    'center',
+    'centered_cutout',
+    'random',
+    'random_max_segments',
+]
+
+
+class Segmenter:
+    """
+    This segmenting returns a list of segmented examples that can be unbatched.
+    Everything that is not listed in `keys` is simply copied from the
+    input example to the output examples.
+    The keys `segment_start` and `segment_stop` are added for each output
+    dictionary.
+
+    If an utterance is shorter than `length`, a
+    `lazy_dataset.FilterException` is raised.
+
+    Examples (For more examples see `tests/test_data/test_segmenter`):
+
+        >>> segmenter = Segmenter(length=32000, include_keys=('x', 'y'),
+        ...                           shift=16000)
+        >>> ex = {'x': np.arange(65000), 'y': np.arange(65000),
+        ...       'num_samples': 65000, 'gender': 'm'}
+        >>> segmented = segmenter(ex)
+        >>> type(segmented)
+        <class 'list'>
+        >>> for entry in segmented:
+        ...     print(entry['x'][[0, -1]])
+        [    0 31999]
+        [16000 47999]
+        [32000 63999]
+
+        Segmenting can be disabled by setting `length=-1`.
+        >>> Segmenter(length=-1, include_keys=('x', 'y'))(ex)[0] == ex
+        True
+
+        Check the corner cases.
+        >>> ex = {'x': np.arange(64000), 'y': np.arange(64000)}
+        >>> for entry in segmenter(ex):
+        ...     print(entry['x'][[0, -1]])
+        [    0 31999]
+        [16000 47999]
+        [32000 63999]
+        >>> ex = {'x': np.arange(63999), 'y': np.arange(63999)}
+        >>> for entry in segmenter(ex):
+        ...     print(entry['x'][[0, -1]])
+        [    0 31999]
+        [16000 47999]
+
+    Args:
+        length: The length of the segments in samples. If set to `-1`,
+            the original example is returned in a list of length one.
```

Comment:
is it possible to add a mode where length is determined automatically based on a max_length argument? E.g. if I have a signal with length 10 and I state max_length=8, then length will be set to 5 so that equally long segments are obtained?

Reply:
You could do that by setting the length and choosing one of the dictionaries from the output

Reply:
I don't think this is what I meant. What I meant is that with a max_length of 8 I, e.g., get
- for a signal of length 10: two segments of length 5
- for a signal of length 15: two segments of length 7.5
- for a signal of length 18: three segments of length 6

Reply:
I guess you mean: Maximize `length <= max_length` so that the input is fully segmented (i.e., there is nothing thrown away). This would certainly complicate the interface and I'm not sure how useful this will be in general. What's the use-case for this? If we decide in favor of this, I'd suggest adding a `length_mode` argument with possible values like "constant" and "max".

Reply:
Isn't this something that should be handled by `get_segment_boundaries` and `segment`? And if we allow this, what about spreading the segments across the input signal (i.e., determine the shift >= length automatically so that the space between segments is equal)? This is in some way a similar feature.